File: lib/ingest/golden_records.ex
# lib/ingest/golden_records.ex — project the re-derived log into golden records (gr-8r6).
#
# Stage 5 of the legacy-medipim ingest, the PRIMARY deliverable, after envelope_loader (gr-n8i),
# claim_mapping (gr-beo) and rederive (gr-chq). Per the design
# (docs/plans/2026-06-05-legacy-history-ingest-design.md, "Ingest pipeline" step 5): fold the
# re-derived log into customer-facing golden records — products ▸ variants ▸ {codes, resolved
# attributes (survivorship), CNK canonical+alias, categories, media} — plus the pass-through
# go-forward `log` so the change feed / lookups stay available downstream.
#
# WHY Catalog.project, NOT History.now (this is the load-bearing choice for this bead) ─────────
# The ingest log carries INTEGER Unix-epoch timestamps (recorded_at / valid_from are ints like
# 1535726805 — see envelope_loader.ex), because the legacy medipim dump is epoch-stamped and the
# ingest does not (yet) convert them. The engine's `History.*` projections AND `Api.get`/`Api.lookup`
# filter via `Date.compare/2`, which RAISES on integers. So we project DIRECTLY through
# `Catalog.project(members, live_claims, priority, overrides)`, which is Date-free: it folds the
# members + the CURRENT claim view, with recency decided by the integer `order` field (stamped by
# ClaimMapping/Rederivation), never by a timestamp comparison. Full temporal projection /
# time-travel over the ingest log is a KNOWN follow-up ("Temporal pass — follow-up, not v1"); it is
# deliberately out of scope here and we do NOT convert timestamps in the merged engine modules.
#
# SAFE engine entrypoints used (all Date-free): `Catalog.project` for the snapshot, and
# `PublicId.canonical(:cnk, key, log, priority)` to enrich each variant's CNK (canonical + aliases).
#
# PRIORITY — a PARAMETER, with a permissive default. The ingest has no canonical medipim
# source-priority table yet (out of scope), so the default `Priority.new(%{}, [])` leaves all
# sources unranked: they tie, and a genuine multi-source disagreement on a field surfaces honestly
# as `status: :needs_review` rather than silently picking a winner. Callers that DO have a ranking
# pass their own `%Priority{}`.
#
# MEDIA arrives via the media LANE (gr-kek): claim_mapping promotes the envelope's "media" and
# "descriptions" references to first-class records in their own lanes, tied to the product by
# :depicts/:describes edges — so `media` carries MED_* records and `descriptions` carries DSC_*
# records with traversal provenance. The legacy :media claim kind still resolves if present.
#
# SCOPE BOUNDARY: this bead produces ONLY the projection. It does NOT build the LegacyXref /
# legacy⟷SK map (gr-0c2), nor the demo script / synthetic fixtures (gr-bxf).

defmodule GoldenRecords do
  # No steward overrides in the PoC — the ingest mints fresh keys and has no human-resolved
  # conflicts to replay, so both override maps are empty.
  @no_overrides %{attr: %{}, product: %{}}

  @doc """
  Project a `%{log, ledger}` re-derivation result (the output of `Rederivation.run/2` or
  `Rederivation.from_claims/2`) into golden records.

  Returns `%{records: [...], log: log}`, where `records` is a list of products, each
  `%{product: <legacy product label>, variants: [variant]}`, and each `variant` is

      %{
        key: "SK_n",                         # surrogate key
        codes: [canonicalized {scheme, value}, ...],
        cnk: %{canonical: {:cnk, "..."}, aliases: [...]} | nil,
        attributes: [{field, %{value, winner, status, candidates}}, ...],  # survivorship-resolved
        product: %{value, winner, status, candidates},
        categories: [{collection, value}, ...],
        media: [%{asset: "MED_n", role, source, uri}, ...],       # via :depicts edges
        substances: [%{key, codes, sources}, ...],                # via :contains edges
        descriptions: [%{key: "DSC_n", via, asserted_by, attributes}, ...]  # derived (gr-sw0)
      }

  `log` is passed through unchanged so downstream callers keep the engine's read layer
  (`Api.resolve_key/2`, `Api.changes_since/2`, `PublicId.collisions/2`).

  `priority` is a `%Priority{}`; it defaults to the permissive `Priority.new(%{}, [])` (see
  moduledoc) so unranked-source conflicts surface as `:needs_review`.
  """
  def project(rederivation, priority \\ default_priority())

  def project(%{log: log, ledger: ledger}, %Priority{} = priority) do
    records =
      ledger.members
      |> Catalog.project(live_claims(log), priority, @no_overrides)
      |> Enum.map(fn %{product: product, variants: variants} ->
        %{product: product, variants: Enum.map(variants, &enrich(&1, log, priority))}
      end)

    %{records: records, log: log}
  end

  @doc """
  Convenience entrypoint: re-derive `envelopes` at instant `at`, then project. Equivalent to
  `envelopes |> Rederivation.run(at) |> project(priority)`.
  """
  def from_envelopes(envelopes, at, priority \\ default_priority()) when is_list(envelopes) do
    envelopes |> Rederivation.run(at) |> project(priority)
  end

  @doc "The permissive default priority — every source unranked, so conflicts tie (see moduledoc)."
  def default_priority, do: Priority.new(%{}, [])

  # Attach the variant's customer-facing CNK (canonical + aliases) from the same log/priority.
  defp enrich(variant, log, priority) do
    Map.put(variant, :cnk, PublicId.canonical(:cnk, variant.key, log, priority))
  end

  # The CURRENT ClaimAsserted view of the log — what Catalog.project folds over.
  defp live_claims(log) do
    log |> Enum.filter(&match?(%Events.ClaimAsserted{}, &1)) |> Substrate.current()
  end
end

File: php/src/GoldenRecords.php
<?php

declare(strict_types=1);

namespace Ingot;

/**
 * Project the re-derived log into golden records — ported from `GoldenRecords`
 * (lib/ingest/golden_records.ex). Folds the members + current claim view via the Date-free
 * `Catalog::project` path, then enriches each variant with its CNK (canonical + aliases).
 */
final class GoldenRecords
{
    /** No steward overrides in the PoC. */
    private const NO_OVERRIDES = ['attr' => [], 'product' => []];

    /**
     * `$priority` is the survivorship policy — a {@see Priority} (tier ranking) OR a
     * `callable(dimension, source): int|float` injected rank fun (medipim's context-aware,
     * off-product-penalty scoring lives there). The toggle is thus reachable from the fold entry,
     * not just {@see Survivorship::decide()}.
     *
     * @param array{log: list<array<string,mixed>>, ledger: LedgerState} $rederivation
     * @return array{records: list<array<string,mixed>>, log: list<array<string,mixed>>}
     */
    public static function project(array $rederivation, Priority|callable|null $priority = null): array
    {
        $priority ??= self::defaultPriority();
        $log = $rederivation['log'];
        $ledger = $rederivation['ledger'];

        $projected = Catalog::project($ledger->members, self::liveClaims($log), $priority, self::NO_OVERRIDES);

        $records = [];
        foreach ($projected as $p) {
            $variants = [];
            foreach ($p['variants'] as $variant) {
                $variants[] = self::enrich($variant, $log, $priority);
            }
            $records[] = ['product' => $p['product'], 'variants' => $variants];
        }

        return ['records' => $records, 'log' => $log];
    }

    /**
     * Convenience: re-derive envelopes at `at`, then project.
     *
     * @param list<array<string,mixed>> $envelopes
     * @return array{records: list<array<string,mixed>>, log: list<array<string,mixed>>}
     */
    public static function fromEnvelopes(array $envelopes, mixed $at, Priority|callable|null $priority = null): array
    {
        return self::project(Rederivation::run($envelopes, $at), $priority ?? self::defaultPriority());
    }

    /** The permissive default priority — every source unranked, so conflicts tie. */
    public static function defaultPriority(): Priority
    {
        return Priority::new([], []);
    }

    /**
     * @param array<string,mixed> $variant
     * @param list<array<string,mixed>> $log
     * @return array<string,mixed>
     */
    private static function enrich(array $variant, array $log, Priority|callable $priority): array
    {
        $variant['cnk'] = PublicId::canonical('cnk', $variant['key'], $log, $priority);

        return $variant;
    }

    /**
     * @param list<array<string,mixed>> $log
     * @return list<array<string,mixed>>
     */
    private static function liveClaims(array $log): array
    {
        $claims = [];
        foreach ($log as $e) {
            if (($e['type'] ?? null) === Events::TYPE_CLAIM_ASSERTED) {
                $claims[] = $e;
            }
        }

        return Substrate::current($claims);
    }
}
